### ISSUE TITLE
Support jwt token auth

### ISSUE THREAD

**Issue by jkroepke** [kiwigrid/k8s-sidecar#321]
As more secure alternative to auth basic, add the support via jwt token.

it should possible to define a token file and a header name to attach the token. The token file can be changed each hour. 

Use case:

i have to disable the initial admin in grafana. I would like to enable jwt auth to allow login via kubernetes server account token, since its a short-living jwt token.

ref: https://github.com/grafana/grafana/issues/81794

**Comment by jkroepke** [comment]
.

**Comment by jkroepke** [comment]
thx for info.

**Comment by jkroepke** [comment]
thx for info, still useful

**Comment by ChristianGeie** [comment]
I myself find the opportunity to use JWT tokens very well. However, I have no idea how to implement it in Python.

Therefore, ideas or pull requests are very welcome here.

**Comment by jkroepke** [comment]
@ChristianGeie basicly it's reading a file (each time, since the content can be changed) and put the content into an Authorization header, e.g. `X-JWT-Assertion: <file-content>`. 

The header name should be configurable, since the header name is [configurable](https://grafana.com/docs/grafana/latest/setup-grafana/configure-security/configure-authentication/jwt/#enable-jwt) in grafana as well.

**Comment by jkroepke** [comment]
> if we use JWT auth do we need basic auth in parallel or is the token auth separat or replace it basic auth complete

I don't know a use case for require both.

> the given token must come from a external identity provider (keycloak, oauth)

Keep in mind that the Kubernetes Service Account Token is a valid JWT by default as well.

> we need to validate ...

I would not validate the JWT. It's the job Grafana instance to validate the JWT token. It's on your choice to add an additional validation in k8s-sidecar, if you like to pull in additional libraries.

**Comment by jkroepke** [comment]
thx for info, still useful

**Comment by jkroepke** [comment]
thx for info, still useful

**Comment by jkroepke** [comment]
thx for info, still useful

**Comment by jkroepke** [comment]
thx for info, still useful

**Comment by jkroepke** [comment]
thx for info, still useful

**Comment by Baarsgaard** [comment]
Hi,

Not too long ago, I spent some time implementing this in the [Grafana-Operator](https://grafana.github.io/grafana-operator/docs/examples/grafana/usekubeauth/#custom-token-audience)!
Some quick notes that might be useful 🙂

> @jkroepke just start to dig a little deeper in python `requests` lib and grafana documentation to collect requirements:
> * header name should be configurable (can be default to `X-JWT-Assertion`)
> * if token validation fails is basic auth the fallback | if not basic auth is in place, API calls should be blocked
> * if we use JWT auth do we need basic auth in parallel or is the token auth separat or replace it basic auth complete

For security reasons, it's probably safer to not provide fallback?

Parallel should also be unnecessary as the JWT contains enough information to establish a signed and verifiable identity.
It's quite normal to see JWTs present in the `Authorization` header e.g. `Authorization: Bearer $TOKEN`
The `Bearer` part often being a requirement

> * the given token must come from a external identity provider (keycloak, oauth)

Depends on the scope, you can implement support for authenticating with an OAuth2 provider, or you can limit it to just Kubernetes ServiceAccounts.
Doing so allows you to read the JWT from a file on the disk, located at `/var/run/secrets/kubernetes.io/serviceaccount/token`

But if this token leaks, any and all permissions assigned to it can be used with the Kubernetes API by whoever obtains it: https://grafana.github.io/grafana-operator/docs/examples/grafana/usekubeauth/#custom-token-audience

We fixed this by requesting an alternative value for the `aud`(audience) claim in a separate token only used in HTTP requests by the [Grafana-Operator](https://github.com/grafana/grafana-operator/pull/2371/changes)
This causes Kubernetes to reject the token.

> * we need to validate the token signature (e.g. against the public key or against the JWKS endpoint of the identity provider)
> * we need to validate the token claims (`aud` Audience or `iss` Issuer) - optional

You can skip validation entirely, it's your "identity" and you don't have any way of fixing an invalid token. It's up to the receiving service to validate your identity against the issuer!
In this case, Grafana will have to validate your identity and the validity of the signature against the jwks kubernetes exposes.

> * we need to validate the token expiry date

The `exp`(expiration) claim in the JWT, it allows you to know how long to cache the token for.
Kubernetes will attempt to replace the projected JWT in the token file after [80%](https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/1205-bound-service-account-tokens/README.md#:~:text=80%20percent) of the tokens lifetime has passed.

Relevant Links:
- https://github.com/grafana/grafana-operator/pull/2137
- https://github.com/grafana/grafana-operator/pull/2371
- https://grafana.github.io/grafana-operator/docs/examples/grafana/usekubeauth/
- https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/
- https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/1205-bound-service-account-tokens/README.md

**Comment by jkroepke** [comment]
@Baarsgaard In summary, it works on Grafana side, but it's still missing on k8s-sidecar?

**Comment by Baarsgaard** [comment]
@jkroepke Yes, Grafana has full support for JWT validation and recently also support for retrieving the JWKS from an authenticated endpoint with a selfsigned Certificate (12.2.0 and on), which was the missing bit for ServiceAccounts.

With regards to the k8s-sidecar, I only know what I could read from this issue and the draft: https://github.com/kiwigrid/k8s-sidecar/pull/388 which I assume means it's not implemented yet.

**Comment by jkroepke** [comment]
Thanks for the update good to know that it's implemented on Grafana side.

**Comment by jkroepke** [comment]
thx for info, still useful

**Comment by jkroepke** [comment]
thx for info, still useful

**Comment by jkroepke** [comment]
thx for info, still useful
